src/transformers/generation/utils.py

```diff
@@ -1939,6 +1939,7 @@ def generate(
         # generating the first new token or not, and we only want to use the embeddings for the first new token)
         if not self.config.is_encoder_decoder and model_input_name == "inputs_embeds":
             model_kwargs["use_cache"] = True
+            generation_config.use_cache = True
```

Comment:
the subsequent step of preparing cache checks generation config, and we miss the step of adding a Cache class. Fails for IDEFICS only because it doesn't prepare cache in "forward" if "use_cache", but rather assume that there's already a correct cache provided

Reply:
This change updates `generation_config`, which means we would have two variables containing the source of truth for this flag (`generation_config.use_cache` and `model_kwargs["use_cache"]`). More than one source of truth usually leads to bugs 🐛 We also know that we need `model_kwargs["use_cache"]` for the forward pass.

To avoid multiple sources of truth, I'd suggest one of the following: 
1. let's use `model_kwargs["use_cache"]` in all places after this if/else, instead of also using `generation_config.use_cache`.
2. (I have a preference for this one) Let's use `generation_config.use_cache` everywhere, removing `model_kwargs["use_cache"]` from most places in the `generate` function. Before calling the decoding methods, let's add `model_kwargs["use_cache"] = generation_config.use_cache`, since we need this for the forward pass and `generation_config` barely gets used in the decoding methods

(I know this issue predates your change 🤗 But since we're touching it, let's do the most sustainable change)